File: 0152-maximum-product-subarray/0152-maximum-product-subarray.java
class Solution { // tc is O(n) and sc is O(1)
    public int maxProduct(int[] nums) {
       
       int max = nums[0],min = nums[0],n = nums.length,result = max;

       for(int i = 1;i<n;i++){
            int curr = nums[i];
            int newMax = Math.max(curr,Math.max(max * curr,min*curr));
            min = Math.min(curr,Math.min(max*curr,min*curr));
            max = newMax;

            result = Math.max(max,result);
       }
       return result;
    }
}

File: 0152-maximum-product-subarray/0152-maximum-product-subarray.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int n = nums.size(),newmin = nums[0],newmax = nums[0],result = newmax;
        
        for(int i = 1;i<n;i++){
            int curr = nums[i];
            
            int newMax = max(curr,max(newmax*curr,newmin*curr));
            newmin = min(curr,min(newmax*curr,newmin*curr));
            newmax = newMax;
            result = max(result,newmax);
        }
        return result;
    }
};